crates/next-core/src/babel.rs

```diff
@@ -0,0 +1,170 @@
+use anyhow::Result;
+use turbo_tasks::{
+    primitives::{BoolVc, StringVc},
+    Value,
+};
+use turbo_tasks_fs::FileSystemPathVc;
+use turbopack::{
+    module_options::WebpackLoadersOptionsVc, resolve_options,
+    resolve_options_context::ResolveOptionsContext,
+};
+use turbopack_core::{
+    issue::{Issue, IssueSeverity, IssueSeverityVc, IssueVc},
+    resolve::{parse::RequestVc, pattern::Pattern, resolve},
+};
+use turbopack_node::transforms::webpack::{WebpackLoaderConfigItem, WebpackLoaderConfigItemsVc};
+
+const BABEL_CONFIG_FILES: &[&str] = &[
+    ".babelrc",
+    ".babelrc.json",
+    ".babelrc.js",
+    ".babelrc.mjs",
+    ".babelrc.cjs",
+    "babel.config.js",
+    "babel.config.json",
+    "babel.config.mjs",
+    "babel.config.cjs",
+];
+
+/// If the user has a babel configuration file (see list above) alongside their
+/// `next.config.js` configuration, automatically add `babel-loader` as a
+/// webpack loader for each eligible file type if it doesn't already exist.
+#[turbo_tasks::function]
+pub async fn maybe_add_babel_loader(
+    project_root: FileSystemPathVc,
+    webpack_options: WebpackLoadersOptionsVc,
+) -> Result<WebpackLoadersOptionsVc> {
+    let has_babel_config = {
+        let mut has_babel_config = false;
+        for filename in BABEL_CONFIG_FILES {
+            let metadata = project_root.join(filename).metadata().await;
+            if metadata.is_ok() {
+                has_babel_config = true;
+                break;
+            }
+        }
+        has_babel_config
+    };
+
+    if has_babel_config {
+        let mut options = (*webpack_options.await?).clone();
+        for ext in [".js", ".jsx", ".ts", ".tsx", ".cjs", ".mjs"] {
```

Comment:
Should this be using `next_config.page_extensions()` instead? Also, what about `.cts`, `.mts`, `.ctsx`, and `.mtsx`?